src/org/traccar/protocol/TeltonikaProtocolDecoder.java

```diff
@@ -233,23 +315,35 @@ private void decodeLocation(Position position, ChannelBuffer buf, int codec) {
         if (BitUtil.check(globalMask, 1)) {
             int cnt = buf.readUnsignedByte();
             for (int j = 0; j < cnt; j++) {
-                decodeParameter(position, buf.readUnsignedByte(), buf, 1);
+                if (codec == CODEC_GH3000) {
+                    decodeGh3000Parameter(position, buf.readUnsignedByte(), buf, 1);
+                } else {
+                    decodeParameter(position, buf.readUnsignedByte(), buf, 1);
+                }
             }
         }
 
         // Read 2 byte data
         if (BitUtil.check(globalMask, 2)) {
             int cnt = buf.readUnsignedByte();
             for (int j = 0; j < cnt; j++) {
-                decodeParameter(position, buf.readUnsignedByte(), buf, 2);
+                if (codec == CODEC_GH3000) {
+                    decodeGh3000Parameter(position, buf.readUnsignedByte(), buf, 2);
+                } else {
+                    decodeParameter(position, buf.readUnsignedByte(), buf, 2);
+                }
             }
         }
 
         // Read 4 byte data
         if (BitUtil.check(globalMask, 3)) {
             int cnt = buf.readUnsignedByte();
             for (int j = 0; j < cnt; j++) {
-                decodeParameter(position, buf.readUnsignedByte(), buf, 4);
+                if (codec == CODEC_GH3000) {
```

Comment:
There is a bit of duplication. I suggest we have 3 methods `decodeGh3000Parameter`, `decodeOtherParameter` and `decodeParameter`. Last one will take `codec` parameter and based on it decide which of the first two methods to call.